modules/library/main/src/test/java/org/geotools/data/store/ReTypingFeatureCollectionTest.java

```diff
@@ -141,4 +144,31 @@ public void testPreserveUserData() throws Exception {
         SimpleFeature first = DataUtilities.first(rtc);
         assertEquals(TEST_VALUE, first.getUserData().get(TEST_KEY));
     }
+
+    @Test
+    public void testSimilarAttributes() throws Exception {
+        SimpleFeatureTypeBuilder stb = new SimpleFeatureTypeBuilder();
+        stb.setName("original");
+        stb.add("bar", Integer.class);
+
+        SimpleFeatureType schema = stb.buildFeatureType();
+        Assert.assertEquals("bar", schema.getDescriptor(0).getType().getName().getLocalPart());
+
+        SimpleFeatureCollection source = new DefaultFeatureCollection("source", schema);
+
+        SimpleFeatureTypeBuilder stb2 = new SimpleFeatureTypeBuilder();
+        stb2.setName("test");
+
+        AttributeTypeBuilder atb = new AttributeTypeBuilder();
+        atb.setName("Integer");
+        atb.setBinding(Integer.class);
+        stb2.add(atb.buildDescriptor("bar"));
+
+        SimpleFeatureType schema2 = stb2.buildFeatureType();
+        Assert.assertEquals("Integer", schema2.getDescriptor(0).getType().getName().getLocalPart());
+        SimpleFeatureCollection retyped = new ReTypingFeatureCollection(source, schema2);
+        try (SimpleFeatureIterator features = retyped.features()) {
```

Comment:
So you have:

- schema *original* with attribute descriptor "bar" / of property type "bar" with java binding Integer
- schema *test* with attribute descriptor "bar" / of property type "Integer" with java binding Integer

So this PR is in effect seeking to ignore the property type; and just focus on if the descriptor name match?

I think what you are proposing would be new functionality; all the ReType methods are really about sub-setting an existing schema (so we can ignore properties that the user is not interested in as part of their query).

Aside:  We *do* have the ability to transform features on the fly here: https://github.com/geotools/geotools/tree/main/modules/extension/transform/src/main/java/org/geotools/data/transform

Reply:
Yes, it was an easy path to overcome a problem I've seen during my usage of GeoTools. Initially I tried with this #4069, but as @aaime pointed out. That was a bit overwhelming.

I still think this was a good proposal, but accept that you don't want it.

Reply:
Actually - I want it ... but it is a significant change that I think we need to go through the community proposal process.

They way I am thinking it is that:

1) retype: handles the common case where Query has been used to shuffle the order of the parameters around; or only select some of the parameters; and expects the java bindings to match. This would in effect be a "fast path" ...

2) covers: (can we think of a better name) relaxes this constraint so it just is based on descriptor name; and expects the java bindings to be assignable from (or even use a converter).

Options on how to engage:
- We could make a query hint to allow "relax=true", or "covers=true", or "strict=false"
- We could try ReType relationship (as a fast path); and if it is not suitable fall back to Covers relationship...

It seems like a useful idea?

Reply:
I see your point when you talk about a "fast path", but looking at the current code I would argue that we're not having a "fast path" at the moment. Within the "next" function there is a usage of SimpleFeatureBuilder, that in turn will call Converters.convert for all attributes of all the features.
I believe a "fast path" would do more like setting value directly from the original attribute into the new attribute. In such a case it would make perfect sense to demand equality among the AttributeType's. At the moment, I don't see the reason to demand this.